### PR TITLE
Pass Opt to TransformerEncoder/Decoder

### PR DESCRIPTION
**Pull request by spencerp**
**Patch description**
Since the RFC was well received (#3470) here's a PR against master (the other PR was branched off of tf-components #3466).

**Testing steps**
Make sure CircleCI passes.

### CODE REVIEW

**Comment by jaseweston** [comment]
....and what will happen with parlai_internal?

**Comment by spencerp** [comment]
> ....and what will happen with parlai_internal?

I'm getting an internal PR ready now!

File: parlai/agents/transformer/modules.py [hunk 1]
```diff
@@ -310,86 +277,58 @@ class TransformerEncoder(nn.Module):
...
     :param int padding_idx: Reserved padding index in the embeddings matrix.
-    :param bool learn_positional_embeddings: If off, sinusoidal embeddings are
-        used. If on, position embeddings are learned from scratch.
+    :param str reduction_type: Type of reduction at the end of the encoder.
+    :param int n_segments: Number of segments/lang/sentence embeddings.
     :param bool embeddings_scale: Scale embeddings relative to their dimensionality.
         Found useful in fairseq.
-    :param bool reduction: If true, returns the mean vector for the entire encoding
-        sequence.
-    :param int n_positions:
```
Comment (jaseweston) [inline parlai/agents/transformer/modules.py:333]: will these comments appear somewhere else then?

**Comment by jaseweston** [comment]
> > ....and what will happen with parlai_internal?
> 
> I'm getting an internal PR ready now!

👍
👍

File: parlai/agents/transformer/modules.py (hunk 1 above)
Comment (spencerp) [reply parlai/agents/transformer/modules.py:333]: In the opt declaration! I just noticed that not all of opt flags have descriptions, though, so I'll fill them in from here where applicable.

File: parlai/agents/transformer/modules.py [hunk 2]
```diff
@@ -310,86 +277,58 @@ class TransformerEncoder(nn.Module):
...
         self.reduction_type = reduction_type
         self.padding_idx = padding_idx
         # this is --dropout, not --relu-dropout or --attention-dropout
-        self.dropout_frac = dropout
+        self.dropout_frac = opt['dropout']
         self.dropout = nn.Dropout(p=self.dropout_frac)
-        self.variant = variant
-        self.n_segments = n_segments
+        self.variant = opt['variant']
+        self.n_segments = n_segments or opt.get('n_segments', 0)
```
Comment (spencerp) [inline parlai/agents/transformer/modules.py:319]: Btw if we find ourselves needing to manually specify any parameters or override what's in opt, we can just add another default None keyword arg and do this.

**Comment by spencerp** [comment]
Hm, Transresnet tests are failing consistently. That's concerning... Will need to get to the bottom of that before merging.

File: parlai/agents/transformer/modules.py (hunk 1 above)
Comment (klshuster) [reply parlai/agents/transformer/modules.py:333]: i suppose that's one thing we lose by getting rid of this, but the net negative lines is so enticing...

File: parlai/agents/transformer/modules.py (hunk 2 above)
Comment (klshuster) [reply parlai/agents/transformer/modules.py:321]: i suppose im a bit curious how you chose which options to keep in the init, and which you left for opt?

File: parlai/agents/transformer/modules.py [hunk 3]
```diff
@@ -658,62 +603,34 @@ class TransformerDecoder(nn.Module):
     """
     Transformer Decoder layer.
```
Comment (klshuster) [inline parlai/agents/transformer/modules.py:604]: minor nit i noticed (unrelated to your patch) is that this docstring is wrong

**Review by klshuster — APPROVED** [review]
I think this looks good - definitely a fan of net negative lines.

I think Jason made a good point about the field/attr descriptions. I like how they're in the opt, but would it also be possible to include a note in the docstring of where to find those descriptions? like, "For more information regarding model parameters, see :file:" or something like that

**Comment by stephenroller** [comment]
Big improvement.

**Pull request merged**